private-identity-audit/src/proofs.rs

```diff
@@ -17,26 +17,150 @@
 
 use crate::errors::{ErrorKind, Fallible};
 use crate::Challenge;
-use cryptography_core::curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
+use cryptography_core::curve25519_dalek::{
+    ristretto::{CompressedRistretto, RistrettoPoint},
+    scalar::Scalar,
+};
 use rand_core::{CryptoRng, RngCore};
 
-#[derive(Clone)]
+use codec::{Decode, Encode, Error as CodecError, Input, Output};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use sp_std::vec::Vec;
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct InitialMessage {
     a: RistrettoPoint,
     pub generators: Vec<RistrettoPoint>,
 }
 
-#[derive(Clone)]
+impl Encode for InitialMessage {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        32 + self.generators.len() * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let a = self.a.compress();
+        let generators: Vec<[u8; 32]> = self
+            .generators
+            .clone()
+            .into_iter()
+            .map(|g| *g.compress().as_bytes())
+            .collect::<Vec<_>>();
+
+        a.as_bytes().encode_to(dest);
```

Comment:
Performance: Remove some intermediate copies.
```suggestion
    let generators = self.generators.iter()
            .map(|g| g.compress().to_bytes())
            .collect::<Vec<_>>();

        self.a.compress().to_bytes().encode_to(dest);
```